tests/trainer/test_trainer.py

```diff
@@ -3769,3 +3769,42 @@ def test_hyperparameter_search_backends(self):
             list(ALL_HYPERPARAMETER_SEARCH_BACKENDS.keys()),
             list(HPSearchBackend),
         )
+
+
+if is_torch_available():
+
+    class OptimizerAndModelInspectionTest(unittest.TestCase):
+        def test_get_num_trainable_parameters(self):
+            model = nn.Sequential(nn.Linear(128, 64), nn.Linear(64, 32))
+            # in_features * out_features + bias
+            layer_1 = 128 * 64 + 64
+            layer_2 = 64 * 32 + 32
+            trainer = Trainer(model=model)
+            self.assertEqual(trainer.get_num_trainable_parameters(), layer_1 + layer_2)
+            # Freeze the last layer
+            for param in model[-1].parameters():
+                param.requires_grad = False
+            self.assertEqual(trainer.get_num_trainable_parameters(), layer_1)
+
+        def test_get_learning_rates(self):
+            model = nn.Sequential(nn.Linear(128, 64))
+            trainer = Trainer(model=model)
+            with self.assertRaises(ValueError):
+                trainer.get_learning_rates()
+            trainer.create_optimizer()
+            self.assertEqual(trainer.get_learning_rates(), [5e-05, 5e-05])
+
+        def test_get_optimizer_group(self):
+            model = nn.Sequential(nn.Linear(128, 64))
+            trainer = Trainer(model=model)
+            # ValueError is raised if optimizer is None
+            with self.assertRaises(ValueError):
+                trainer.get_optimizer_group()
+            trainer.create_optimizer()
+            # Get groups
+            num_groups = len(trainer.get_optimizer_group())
+            self.assertEqual(num_groups, 2)
+            # Get group of parameter
+            param = next(model.parameters())
+            group = trainer.get_optimizer_group(param)
+            self.assertIn(param, group["params"])
```

Comment:
We use `@require_xxx` decorators to control this. This makes it possible to add several requirements without having many indents

```suggestion
@require_torch
class OptimizerAndModelInspectionTest(unittest.TestCase):
    def test_get_num_trainable_parameters(self):
        model = nn.Sequential(nn.Linear(128, 64), nn.Linear(64, 32))
        # in_features * out_features + bias
        layer_1 = 128 * 64 + 64
        layer_2 = 64 * 32 + 32
        trainer = Trainer(model=model)
        self.assertEqual(trainer.get_num_trainable_parameters(), layer_1 + layer_2)
        # Freeze the last layer
        for param in model[-1].parameters():
            param.requires_grad = False
        self.assertEqual(trainer.get_num_trainable_parameters(), layer_1)

    def test_get_learning_rates(self):
        model = nn.Sequential(nn.Linear(128, 64))
        trainer = Trainer(model=model)
        with self.assertRaises(ValueError):
            trainer.get_learning_rates()
        trainer.create_optimizer()
        self.assertEqual(trainer.get_learning_rates(), [5e-05, 5e-05])

        def test_get_optimizer_group(self):
            model = nn.Sequential(nn.Linear(128, 64))
            trainer = Trainer(model=model)
            # ValueError is raised if optimizer is None
            with self.assertRaises(ValueError):
                trainer.get_optimizer_group()
            trainer.create_optimizer()
            # Get groups
            num_groups = len(trainer.get_optimizer_group())
            self.assertEqual(num_groups, 2)
            # Get group of parameter
            param = next(model.parameters())
            group = trainer.get_optimizer_group(param)
            self.assertIn(param, group["params"])
```

Reply:
Great, thank you very much. That makes perfect sense :)